ethereum/contractBackend.go

```diff
@@ -0,0 +1,97 @@
+// Use this function ONLY IN TEST PACKAGES to set a simulated backend and test wallets with weak encryption params.
+func InitCustomContractBackend(cb *ethchannel.ContractBackend) *ContractBackend {
+	return &ContractBackend{cb: cb}
+}
+
+// InitContractBackend initializes a connection to blockchain node and
+// sets up a wallet for funding onchain transactions.
+func InitContractBackend(url, userKeystorePath string, userAddr wallet.Address) (_ *ContractBackend, err error) {
+	var ethereumBackend *ethclient.Client
+	ethereumBackend, err = ethclient.Dial(url)
+	if err != nil {
+		return nil, errors.WithMessage(err, "connecting to ethereum node")
+	}
+	ks := keystore.NewKeyStore(userKeystorePath, standardScryptN, standardScryptP)
+	acc := &accounts.Account{Address: common.BytesToAddress(userAddr.Bytes())}
+	cb := ethchannel.NewContractBackend(ethereumBackend, ks, acc)
+	return InitCustomContractBackend(&cb), nil
+}
+
+// NewFunder initializes and returns an instance of ethereum funder.
+func (cb *ContractBackend) NewFunder(assetAddr wallet.Address) channel.Funder {
+	return newFunder(*cb.cb, assetAddr)
+}
+
+// NewAdjudicator initializes and returns an instance of ethereum adjudicator.
+func (cb *ContractBackend) NewAdjudicator(adjAddr, receiverAddr wallet.Address) channel.Adjudicator {
+	return newAdjudicator(*cb.cb, adjAddr, receiverAddr)
+}
+
+// ValidateContracts validates the integrity of given adjudicator and asset holder contracts.
+func (cb *ContractBackend) ValidateContracts(adjAddr, assetAddr wallet.Address) error {
+	return validateContracts(*cb.cb, adjAddr, assetAddr)
+}
+
+// DeployAdjudicator deploys the adjudicator contract.
+func (cb *ContractBackend) DeployAdjudicator(ctx context.Context) (wallet.Address, error) {
+	return deployAdjudicator(ctx, *cb.cb)
+}
+
+// DeployAsset deploys the asset holder contract, setting the adjudicator address to given value.
+func (cb *ContractBackend) DeployAsset(ctx context.Context, adjAddr wallet.Address) (wallet.Address, error) {
+	return deployAsset(ctx, *cb.cb, adjAddr)
+}
+
+func newFunder(cb ethchannel.ContractBackend, assetAddr wallet.Address) channel.Funder {
+	return ethchannel.NewETHFunder(cb, ethwallet.AsEthAddr(assetAddr))
+}
+
+func newAdjudicator(cb ethchannel.ContractBackend, adjAddr, receiverAddr wallet.Address) channel.Adjudicator {
+	return ethchannel.NewAdjudicator(cb, ethwallet.AsEthAddr(adjAddr), ethwallet.AsEthAddr(receiverAddr))
+}
+
+func validateContracts(cb ethchannel.ContractBackend, adjAddr, assetAddr wallet.Address) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	// Integrity of Adjudicator is implicitly done during validation of asset holder contract.
+	err := ethchannel.ValidateAssetHolderETH(ctx, cb, ethwallet.AsEthAddr(assetAddr), ethwallet.AsEthAddr(adjAddr))
+	if ethchannel.IsContractBytecodeError(err) {
+		return errors.New("invalid contracts at given address")
```

Comment:
I think somewhere above you used `errors.Wrap`, so either use `Wrap` everywhere for errors where you dont know if they are already wrapped or not, or always use `WithMessage`.
In this line either `Wrap` or `WithMessage` since the original cause is lost otherwise.  
EDIT: Since you dont handle `ContractBytecodeError` errors differently, you dont need the `if`.

Reply:
Replaces `WithMessage` with `Wrap` when the stack trace is not yet captured.
I use a different error message for `ContractByteCodeError`, to differentiate between invalid byte code and any error in the process of validating the byte code.